[base-ui] "useId" import error after updating esbuild

### Steps to reproduce

Link to live example: [CodeSandbox](https://codesandbox.io/p/github/patrykszwed/mui-repro-lib/main?import=true&layout=%257B%2522sidebarPanel%2522%253A%2522GIT%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clst5y27l00063b6k99408kjn%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clst5y27l00023b6khwnv8u40%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clst5y27l00043b6k2dy901c3%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clst5y27l00053b6kjq3578tq%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clst5y27l00023b6khwnv8u40%2522%253A%257B%2522id%2522%253A%2522clst5y27l00023b6khwnv8u40%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clst5y27l00053b6kjq3578tq%2522%253A%257B%2522id%2522%253A%2522clst5y27l00053b6kjq3578tq%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clst5y27l00043b6k2dy901c3%2522%253A%257B%2522id%2522%253A%2522clst5y27l00043b6k2dy901c3%2522%252C%2522activeTabId%2522%253A%2522clst5y27l00033b6k9fhs8g4k%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clst5y27l00033b6k9fhs8g4k%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clst5zvqh0019dcgic5d68exw%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522CSB_RUN_OUTSIDE_CONTAINER%253D1%2520devcontainer%2520templates%2520apply%2520--template-id%2520%255C%2522ghcr.io%252Fdevcontainers%252Ftemplates%252Ftypescript-node%255C%2522%2520--template-args%2520%27%257B%257D%27%2520--features%2520%27%255B%255D%27%2522%252C%2522id%2522%253A%2522clst5z03600673b6kd2fsi4v6%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

Steps:
1. Run `npm run build`.
2. Open `/lib/cjs/index.csj.js` and `/lib/esm/index.js` files.
3. Search for `maybeReactUseId` variable in both of them. 


### Current behavior

**Note:** this issue is similar to #37182, #37183 and #29860. Its root cause is probably related to that [Webpack issue](https://github.com/webpack/webpack/issues/14814).

### Observations

What can be observed in `/lib/cjs/index.cjs.js` file:

```
var maybeReactUseId = React6["useId".toString()];
```

What can be observed in `/lib/esm/index.js` file:

```
var maybeReactUseId = React6.useId;
```

That difference is caused by enabling `minifySyntax: true` in `bundle.js` only for the "esm" format. `esbuild` was not minifying `["useId".toString()]` to `.useId` until version `0.18.20` (including). However, it's been doing this since `0.19.0` release (I've checked the latest `esbuild` version and the outcome is the same).

### Scenario

Now let's say this package is an internal components library which is reused in another app that also uses React 17. When one tries to build that app using `webpack` an error will be thrown:

```
export 'useId' (imported as 'X') was not found in 'react' (possible exports: Children, Component, Fragment, Profiler, PureComponent, StrictMode, Suspense, __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED, cloneElement, createContext, createElement, createFactory, createRef, forwardRef, isValidElement, lazy, memo, useCallback, useContext, useDebugValue, useEffect, useImperativeHandle, useLayoutEffect, useMemo, useReducer, useRef, useState, version)
```

I've encountered this issue after updating Vite, in that internal components library, from version 4.4.12 to 5.1.3. Vite uses `esbuild` for client builds minification by default, and in 5.0 they bumped its required `esbuild` version from `^0.18.0` to `^0.19.0`. This is how the `useId` hook is called in the output code in that component library: 

![image](https://github.com/mui/material-ui/assets/32751479/ca2b8d15-6183-4f3e-804c-9b7c55583c23)

As a result, when I'm trying to build my main app (that uses the components library) using `webpack`, the build fails.

### Expected behavior

Bundled `maybeReactUseId` variable (or any other mangled name), should probably use a bracket notation even when `minifySyntax` is enabled on `esbuild@0.19.0` (or newer).

### Context

I have two packages:
- components library:
1. react `17.0.2`
2. @mui/material `5.15.10`
3. vite `5.1.3`
4. esbuild `0.19.12`

- a main app that uses the above components library as a dependency:
1. react `17.0.2`
2. webpack `5.90.3`

When I'm trying to build the main app using `webpack`, I'm getting the error described above. It is possible to work around that by setting `minifySyntax: false` in `esbuild` settings but it'd be great that wouldn't turn out to be the only way.

### Your environment

<details>
  <summary><code>npx @mui/envinfo</code></summary>
A browser doesn't matter in that case.

```
  System:
    OS: macOS 14.3.1
  Binaries:
    Node: 18.18.2 - ~/.nvm/versions/node/v18.18.2/bin/node
    npm: 9.8.1 - ~/.nvm/versions/node/v18.18.2/bin/npm
    pnpm: Not Found
  Browsers:
    Chrome: 121.0.6167.184
    Edge: Not Found
    Safari: 17.3.1
  npmPackages:
    @emotion/react: ^11.11.0 => 11.11.0 
    @emotion/styled: ^11.11.0 => 11.11.0 
    @mui/base:  5.0.0-beta.36 
    @mui/core-downloads-tracker:  5.15.10 
    @mui/material: ^5.15.10 => 5.15.10 
    @mui/private-theming:  5.15.9 
    @mui/styled-engine:  5.15.9 
    @mui/system:  5.15.9 
    @mui/types:  7.2.13 
    @mui/utils:  5.15.9 
    @types/react: ^17.0.0 => 17.0.59 
    react: ^17.0.0 => 17.0.2 
    react-dom: ^17.0.0 => 17.0.2 
    typescript: ^5.0.4 => 5.0.4 
```
</details>


**Search keywords**: esbuild useId utils webpack vite react

@brijeshb42 our setup is pretty wild and we're using `yarn` in the app repo while using `npm` in the components package. When working locally, we're using `link`s to link the local package version. 

Juggling between different webpack and rollup configurations results, in most cases, in either:
1. `Invalid hook call.` error (multiple `react` versions detected). We solve this by specifying an alias for `react` in our webpack config, that is: `alias: { react: require.resolve('react') }`
2. Or when defining `@mui/material` as an external dependency - `Module not found: Error: Can't resolve 'react/jsx-runtime' in '{MY_LOCAL_PATH}/node_modules/@mui/system/esm/cssVars'`. This one can be solved by removing an `alias` for `react`, but once that's done we're going back to the first issue. We've been trying to fix that by using webpack `fallback` and/or `alias` options but without luck so far.

--- a/packages/mui-system/src/useMediaQuery/useMediaQuery.ts
+++ b/packages/mui-system/src/useMediaQuery/useMediaQuery.ts
@@ -72,7 +72,8 @@ function useMediaQueryOld(
   return match;
 }
 
-// eslint-disable-next-line no-underscore-dangle, @typescript-eslint/naming-convention -- Workaround for https://github.com/webpack/webpack/issues/14814
+// See https://github.com/mui/material-ui/issues/41190#issuecomment-2040873379 for why
+// eslint-disable-next-line no-underscore-dangle, @typescript-eslint/naming-convention
 const safeReact = { ...React };
 const maybeReactUseSyncExternalStore: undefined | any = safeReact.useSyncExternalStore;
 

--- a/packages/mui-system/src/useMediaQuery/useMediaQuery.ts
+++ b/packages/mui-system/src/useMediaQuery/useMediaQuery.ts
@@ -73,7 +73,6 @@ function useMediaQueryOld(
 }
 
 // See https://github.com/mui/material-ui/issues/41190#issuecomment-2040873379 for why
-// eslint-disable-next-line no-underscore-dangle, @typescript-eslint/naming-convention
 const safeReact = { ...React };
 const maybeReactUseSyncExternalStore: undefined | any = safeReact.useSyncExternalStore;
 

--- a/packages/mui-system/src/useMediaQuery/useMediaQuery.ts
+++ b/packages/mui-system/src/useMediaQuery/useMediaQuery.ts
@@ -71,8 +71,9 @@ function useMediaQueryOld(
   return match;
 }
 
-// eslint-disable-next-line no-useless-concat -- Workaround for https://github.com/webpack/webpack/issues/14814
-const maybeReactUseSyncExternalStore: undefined | any = (React as any)['useSyncExternalStore' + ''];
+// eslint-disable-next-line no-underscore-dangle, @typescript-eslint/naming-convention -- Workaround for https://github.com/webpack/webpack/issues/14814
+const _React = { ...React };
+const maybeReactUseSyncExternalStore: undefined | any = _React.useSyncExternalStore;
 
 function useMediaQueryNew(
   query: string,

--- a/packages/mui-system/src/useMediaQuery/useMediaQuery.ts
+++ b/packages/mui-system/src/useMediaQuery/useMediaQuery.ts
@@ -30,6 +30,7 @@ export interface UseMediaQueryOptions {
   ssrMatchMedia?: (query: string) => { matches: boolean };
 }
 
+// TODO React 17: Remove `useMediaQueryOld` once React 17 support is removed
 function useMediaQueryOld(
   query: string,
   defaultMatches: boolean,
@@ -149,7 +150,6 @@ export default function useMediaQuery<Theme = unknown>(
   let query = typeof queryInput === 'function' ? queryInput(theme) : queryInput;
   query = query.replace(/^@media( ?)/m, '');
 
-  // TODO: Drop `useMediaQueryOld` and use  `use-sync-external-store` shim in `useMediaQueryNew` once the package is stable
   const useMediaQueryImplementation =
     maybeReactUseSyncExternalStore !== undefined ? useMediaQueryNew : useMediaQueryOld;
   const match = useMediaQueryImplementation(

--- a/packages/mui-system/src/useMediaQuery/useMediaQuery.ts
+++ b/packages/mui-system/src/useMediaQuery/useMediaQuery.ts
@@ -73,8 +73,8 @@ function useMediaQueryOld(
 }
 
 // eslint-disable-next-line no-underscore-dangle, @typescript-eslint/naming-convention -- Workaround for https://github.com/webpack/webpack/issues/14814
-const _React = { ...React };
-const maybeReactUseSyncExternalStore: undefined | any = _React.useSyncExternalStore;
+const safeReact = { ...React };
+const maybeReactUseSyncExternalStore: undefined | any = safeReact.useSyncExternalStore;
 
 function useMediaQueryNew(
   query: string,

--- a/packages/mui-utils/src/useId/useId.test.js
+++ b/packages/mui-utils/src/useId/useId.test.js
@@ -1,7 +1,7 @@
 import * as React from 'react';
 import { expect } from 'chai';
 import { createRenderer, screen } from '@mui/internal-test-utils';
-import useId from './useId';
+import useId from '@mui/utils/useId';
 
 describe('useId', () => {
   const { render, renderToString } = createRenderer();

--- a/packages/mui-utils/src/useId/useId.ts
+++ b/packages/mui-utils/src/useId/useId.ts
@@ -20,7 +20,8 @@ function useGlobalId(idOverride?: string): string | undefined {
   return id;
 }
 
-// eslint-disable-next-line no-underscore-dangle, @typescript-eslint/naming-convention -- Workaround for https://github.com/webpack/webpack/issues/14814
+// See https://github.com/mui/material-ui/issues/41190#issuecomment-2040873379 for why
+// eslint-disable-next-line no-underscore-dangle, @typescript-eslint/naming-convention
 const safeReact = { ...React };
 const maybeReactUseId: undefined | (() => string) = safeReact.useId;
 

--- a/packages/mui-utils/src/useId/useId.ts
+++ b/packages/mui-utils/src/useId/useId.ts
@@ -2,6 +2,8 @@
 import * as React from 'react';
 
 let globalId = 0;
+
+// TODO React 17: Remove `useGlobalId` once React 17 support is removed
 function useGlobalId(idOverride?: string): string | undefined {
   const [defaultId, setDefaultId] = React.useState(idOverride);
   const id = idOverride || defaultId;

--- a/packages/mui-utils/src/useId/useId.ts
+++ b/packages/mui-utils/src/useId/useId.ts
@@ -21,6 +21,7 @@ function useGlobalId(idOverride?: string): string | undefined {
 // eslint-disable-next-line no-underscore-dangle, @typescript-eslint/naming-convention -- Workaround for https://github.com/webpack/webpack/issues/14814
 const _React = { ...React };
 const maybeReactUseId: undefined | (() => string) = _React.useId;
+
 /**
  *
  * @example <div id={useId()} />
@@ -28,10 +29,12 @@ const maybeReactUseId: undefined | (() => string) = _React.useId;
  * @returns {string}
  */
 export default function useId(idOverride?: string): string | undefined {
+  // React.useId() is only available from React 17.0.0.
   if (maybeReactUseId !== undefined) {
     const reactId = maybeReactUseId();
     return idOverride ?? reactId;
   }
+
   // TODO: uncomment once we enable eslint-plugin-react-compiler // eslint-disable-next-line react-compiler/react-compiler
   // eslint-disable-next-line react-hooks/rules-of-hooks -- `React.useId` is invariant at runtime.
   return useGlobalId(idOverride);

--- a/packages/mui-utils/src/useId/useId.ts
+++ b/packages/mui-utils/src/useId/useId.ts
@@ -21,8 +21,8 @@ function useGlobalId(idOverride?: string): string | undefined {
 }
 
 // eslint-disable-next-line no-underscore-dangle, @typescript-eslint/naming-convention -- Workaround for https://github.com/webpack/webpack/issues/14814
-const _React = { ...React };
-const maybeReactUseId: undefined | (() => string) = _React.useId;
+const safeReact = { ...React };
+const maybeReactUseId: undefined | (() => string) = safeReact.useId;
 
 /**
  *

--- a/packages/mui-utils/src/useId/useId.ts
+++ b/packages/mui-utils/src/useId/useId.ts
@@ -18,8 +18,9 @@ function useGlobalId(idOverride?: string): string | undefined {
   return id;
 }
 
-// downstream bundlers may remove unnecessary concatenation, but won't remove toString call -- Workaround for https://github.com/webpack/webpack/issues/14814
-const maybeReactUseId: undefined | (() => string) = (React as any)['useId'.toString()];
+// eslint-disable-next-line no-underscore-dangle, @typescript-eslint/naming-convention -- Workaround for https://github.com/webpack/webpack/issues/14814
+const _React = { ...React };
+const maybeReactUseId: undefined | (() => string) = _React.useId;
 /**
  *
  * @example <div id={useId()} />

--- a/packages/mui-utils/src/useId/useId.ts
+++ b/packages/mui-utils/src/useId/useId.ts
@@ -21,7 +21,6 @@ function useGlobalId(idOverride?: string): string | undefined {
 }
 
 // See https://github.com/mui/material-ui/issues/41190#issuecomment-2040873379 for why
-// eslint-disable-next-line no-underscore-dangle, @typescript-eslint/naming-convention
 const safeReact = { ...React };
 const maybeReactUseId: undefined | (() => string) = safeReact.useId;
 

